Istio HA - HPAs

stepanstipl opened the pull request:
This PR adds management of Istio HPAs to make Istio deployment Highly Available. All the components are set to min 2 instances (while still allowing them to autoscale up to 5 - GKE's default), except for `istio-ingressgateway` which is set to 3 instances as this is component in the critical path accessed externally via the load-balancer and more instances give the gcp load-balancer health checks a bit more time to settle in situations like node auto-upgrade.

Unfortunately, as Istio components are created and managed by Google (via Kubernetes add-on manager), they have to be imported to Terraform state first, before being modified via Terraform.

natarajaya commented on gcp/modules/istio/hpa.tf:
```diff
@@ -0,0 +1,119 @@
+      api_version = "apps/v1beta1"
+    }
+
+    target_cpu_utilization_percentage = 80
```
How did you determine this number?

stepanstipl replied:
This is an existing configuration by Google - ie. it does not change anything, but has to be here, otherwise TF would try to remove it from the rule.

natarajaya commented on shared/rakefiles/xk.rake:
```diff
@@ -17,6 +17,23 @@ task :xk, [:cmd, :skip_secret_mgmt, :preserve_stderr] => [:configure, :configure
 
   Rake::Task["set_secrets"].invoke
 
+  sh_filter "sh -c '
```
If I understand things correctly and this is permanent, I don't think `xk.rake` is a good place for this code to live, since it is going to be called on every `sh` and `xk` runs. Maybe it should be a separate resource of `istio` module and HPA resources should depend on it?

stepanstipl replied:
It has to be run every time, because if you delete that resource (`rake destroy_module` for example), GKE will recreate it. So next time you need to import it again, otherwise TF would fail with an error that the resource already exists.

> it should be a separate resource of istio module

It can't, as the TF state of that module has to be modified before that module plan is executed. Once the module is executed it's too late to modify its state.

I don't like it very much here, but given it has to be called every time before `istio` module runs,  I don't have a better place for it - open to suggestions.

natarajaya replied:
How about moving this to a new `istio-import` module so it can be then added to dependencies of `istio`?

stepanstipl replied:
Well, first it seems conceptually very ugly to have one module modify the state of another.

Apart from that - how would you actually do that? Have a local-exec calling terragrunt? So end up with  `terragrunt` executing `terraform` executing `terragrunt` executing `terrafrom`?.  Also I believe terragrunt executes modules in some weird tmp directories, so you would probably have to hardcode/pass paths where the code/istio module is actually located inside the container originally?

Would this work when calling `deploy_module`? (I don't think so - calling `deploy_module[k8s/gpii/flowmanager]` does not run couchdb module & its `finish_cluster` once it has been deployed).

I was thinking about putting this into a module, but for the above reasons I don't think is a good place. Another alternative I was thinking about is to create a separate task for this, but that would actually make it more expensive.

Otherwise, the cost of this once the import happens, is basically one `terraform state pull` which is imho not too bad (also it shows another issue with our current code structure).

mrtyler replied:
> (also it shows another issue with our current code structure)

A bit off-topic but how would you solve this problem in the theoretical "pure terraform" version of gpii-infra?

stepanstipl replied:
> As written, will this code work under common/ where @env/k8s/istio doesn't exist?

Yes it would - I've explained in the comment about checking return code.

> I think there are many tasks that don't care if the tfstate for istio/ is synced with the reality in GKE: sh, any deploy that doesn't touch istio/, any destroy, numerous utility tasks. Am I missing a connection?

Yes you're correct. The response was in the context it's not sufficient to run it only once. To be more precise the code should run every time before task that either directly or indirectly triggers plan or apply of `istio` module.

> Move this code to its own task in xk_configure.rake
> Optionally invoke the new task from :xk, a la skip_secret_mgmt (but I would make it default to off and call it sync_gke_istio_state or something)
> Add the sync_gke_istio_state flag only where it is needed (deploy* mostly)
> ?

Yep, I like this, looks better than the current state.

> A bit off-topic but how would you solve this problem in the theoretical "pure terraform" version of gpii-infra?

It would be easier to implement as there would be only a single entry point for all plan/apply runs (notice we're not even using native rake dependency mechanism, but implementing parametrized logic instead) and also the redundant runs would be avoided.

stepanstipl commented on shared/rakefiles/xk.rake:
```diff
@@ -17,6 +17,23 @@ task :xk, [:cmd, :skip_secret_mgmt, :preserve_stderr] => [:configure, :configure
+  sh_filter "sh -c '
+    # retrieve TF state
+    RESOURCES=\"$(terragrunt state pull --terragrunt-working-dir \"live/#{@env}/k8s/istio\" | jq -er \".modules[].resources\")\"
+    [ \"$?\" -ne 0 ] && exit 1
```
Yes, this fails only in the case that `terragrunt` is actually able to retrieve the state but `jq` is not able to find `.modules[].resources`, which is fine because if the state is completely empty it will be skipped in the next step. It's the "weird" behaviour of jq's  `-e` flag which works. It also works well in case when the state does not exist, because it has been deleted or it's the first time we run. In this case `terragrunt pull` creates the state and the `RESOURCES` is set to `{}`, which will correctly trigger the import.

But I agree it's a bit confusing and I'll try to simplify or at least explain the logic in the refactor.

stepanstipl commented on gcp/modules/istio/hpa.tf:
```diff
@@ -0,0 +1,119 @@
+  }
+
+  spec {
+    max_replicas = 5
```
Done in the refactor

stepanstipl commented:
> This seems pretty weird, so I'd like to see some high-level comments explaining: what it does
why it is needed. Basically, what you wrote in the PR but a little closer to the code so that readers can make sense of what they're looking at. I would keep the documentation about magic numbers (80%, 5 max) close to the config so there's a chance that someone would remember to update the comment while updating the value.

Yep, makes sense.

> I don't want to add too much scope/time to this task since we are still in FERPA mode, but perhaps this is an opportunity to illustrate how to call a standalone shell script rather than using inline shell (the "API" looks simple -- @env, the live directory structure, and kubectl config)? Or to implement this with (unit-tested) ruby?

While I still have some objections, I see that making the shell code external so that it's lintable, without crazy escaping and with well-defined interface brings value, so 👍 


Thanks everyone for the reviews, I'm refactoring the code as per replies to individual comments:
- create a dedicated task, which is triggered based on xk's parameter
- move actual shell code to an external script
- include comments explaining this and "magic" values in the code
- maybe simplify and/or explain the error handling logic/scenarios
  (note to myself - common, no state, pull fail)

stepanstipl commented:
I believe I've addressed all the concerns raised, please have a look. Changes made are mainly:
- Created a dedicated task, which is triggered based on xk task' parameter
- Move actual shell code to an external script
- Included comments explaining why is the module/task required and "magic" values
- Simplified the script logic and va minor improvements (incl. making script pass shellcheck)
- Added the import task to `gcp-plan*` CI steps - (pls. let me know if you have idea for a nicer way)

I've tested the following scenarios
  - running `rake` and `rake deploy_module[k8s/istio]` repeatedly in dev env
  - running `deploy_module[k8s/istio]` after `destroy_module[k8s/istio]` and also after `destroy_state[k8s/istio]`
  - running `rake` from scratch
  - running `rake 'deploy_module...` under `/common`

mrtyler commented on shared/rakefiles/scripts/sync_gke_istio_state.sh:
```diff
@@ -0,0 +1,62 @@
+# check the module does exist, otherwise silently exit
+if [ ! -d "${ISTIO_MODULE_DIR}" ]
+then
+  echo "${THIS_SCRIPT}: Istio module not found (${ISTIO_MODULE_DIR})"
```
This message looks like an error. Can you clarify that this situation is not (necessarily) bad?

stepanstipl replied:
👍, done
